File: algorithm/src/main/java/baekjoon/p02747/Main.java
package baekjoon.p02747;

import java.util.Arrays;
import java.util.Scanner;

public class Main {

    private static int[] cache;

    public static void main(String[] args) {
        cache = new int[46];
        Arrays.fill(cache, -1);

        Scanner sc = new Scanner(System.in);
//        System.out.println(solve(sc.nextInt()));
        System.out.println(solve2(sc.nextInt()));
    }

    private static int solve(int n) {
        if (n <= 1) {
            return n;
        }

        if (cache[n] != -1) {
            return cache[n];
        }

        int ret = solve(n - 2) + solve(n - 1);
        cache[n] = ret;

        return ret;
    }

    private static int solve2(int n) {
        int[] ret = new int[n + 1];
        ret[0] = 0;
        ret[1] = 1;

        for (int i = 2; i <= n; i++) {
            ret[i] = ret[i - 2] + ret[i - 1];
        }

        return ret[n];
    }

}

File: algorithm/src/main/java/baekjoon/p02747/Main.kt
package baekjoon.p02747

fun main(args: Array<String>) {
    val n = readLine()!!.toInt()
//    println(solve(n))
    println(solve2(n))
}

private val cache = IntArray(46) { -1 }

private fun solve(n: Int): Int {
    if (n <= 1) {
        return n
    }

    if (cache[n] != -1) {
        return cache[n]
    }

    val ret = solve(n - 2) + solve(n - 1)
    cache[n] = ret

    return ret
}

private fun solve2(n: Int): Int {
    val ret = IntArray(46)
    ret[0] = 0
    ret[1] = 1

    for (i in 2..n) {
        ret[i] = ret[i - 2] + ret[i - 1]
    }

    return ret[n]
}